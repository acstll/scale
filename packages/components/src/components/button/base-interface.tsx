import { CssClassMap } from "../../utils/utils";

export default interface Base {
	theme: any;
	stylesheet: any;
	componentWillLoad(): void;
	getCssClassMap(): CssClassMap;
}
