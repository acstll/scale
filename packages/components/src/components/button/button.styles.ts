import { JssStyle } from 'jss';

const styles: JssStyle = {
  button: {
    position: 'relative',
    display: 'inline-flex',
    justifyContent: 'center',
    alignItems: 'center',
    flexShrink: 0,
    verticalAlign: 'middle',
    textAlign: 'center',
    cursor: 'pointer',
    userSelect: 'none',
    color: '#333',
    background: '#fff',
    border: '1px solid #000',
    padding: '0.25rem 1rem',
    lineHeight: '2rem',
    transition: 'all 0.2s ease-in-out',
    fontSize: '1rem',
    '&:hover, &.active': {
      color: '#fff',
      background: '#000',
      borderColor: '#000',
      transition: 'all 0.2s ease-in-out',
      textDecoration: 'none',
    },
    '&:not(.tabbing):focus': {
      outline: 0,
    },
    '&--disabled, &--disabled:hover': {
      background: '#fff',
      border: '1px solid #000',
      color: '#000',
      opacity: 0.5,
      cursor: 'not-allowed',
    },
  },
};

export default styles;
