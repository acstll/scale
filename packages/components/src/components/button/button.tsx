import { Component, Prop, h, Method, Host } from '@stencil/core';
import { StyleSheet } from 'jss';
import classNames from 'classnames';
import { CssClassMap } from '../../utils/utils';
import { CssInJs } from '../../utils/decorators/css-in-js';
import Base from '../../utils/base-interface';
import defaultStyles from './button.styles';

@Component({
  tag: 't-button',
  shadow: true,
})
export class Button implements Base {
  /** (optional) Button size */
  @Prop() size?: string = '';
  /** (optional) Button variant */
  @Prop() variant?: string = '';
  /** (optional) Disabled button */
  @Prop() disabled?: boolean = false;
  /** (optional) Deselected button */
  @Prop() deselected?: boolean = false;

  /** (optional) Injected jss styles */
  @Prop() styles?: StyleSheet;
  /** decorator Jss stylesheet */
  @CssInJs('Button', defaultStyles) stylesheet: StyleSheet;

  /** method: componentWillLoad()  */
  componentWillLoad() {}

  /** method: disable()  */
  @Method()
  async disable() {
    this.disabled = true;
  }

  /** method: enable()  */
  @Method()
  async enable() {
    this.disabled = false;
  }

  render() {
    return (
      <Host>
        <style>{this.stylesheet.toString()}</style>
        <button class={this.getCssClassMap()} disabled={this.disabled}>
          <slot />
        </button>
      </Host>
    );
  }

  getCssClassMap(): CssClassMap {
    const { classes } = this.stylesheet;
    return classNames(
      classes.button,
      this.size && `${classes.button}--size-${this.size}`,
      this.variant && `${classes.button}--variant-${this.variant}`,
      this.disabled && `${classes.button}--disabled`,
      this.deselected && `${classes.button}--deselected`
    );
  }
}
