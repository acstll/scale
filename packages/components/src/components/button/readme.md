# t-button



<!-- Auto Generated Below -->


## Properties

| Property     | Attribute    | Description                    | Type                                     | Default     |
| ------------ | ------------ | ------------------------------ | ---------------------------------------- | ----------- |
| `deselected` | `deselected` | (optional) Deselected button   | `boolean`                                | `false`     |
| `disabled`   | `disabled`   | (optional) Disabled button     | `boolean`                                | `false`     |
| `size`       | `size`       | (optional) Button size         | `string`                                 | `''`        |
| `styles`     | --           | (optional) Injected jss styles | `StyleSheet<string \| number \| symbol>` | `undefined` |
| `theme`      | `theme`      |                                | `any`                                    | `undefined` |
| `variant`    | `variant`    | (optional) Button variant      | `string`                                 | `''`        |


## Methods

### `disable() => Promise<void>`

Button method: disable()

#### Returns

Type: `Promise<void>`



### `enable() => Promise<void>`

Button method: enable()

#### Returns

Type: `Promise<void>`




----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
