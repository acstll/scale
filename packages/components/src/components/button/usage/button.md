<t-button>
	Label
</t-button>
