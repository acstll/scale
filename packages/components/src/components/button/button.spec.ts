import { newSpecPage } from '@stencil/core/testing';
import { Button } from './button';
import jss from 'jss';
import styles from './button.styles';

describe('Button', () => {
  let element;
  beforeEach(async () => {
    element = new Button();
    element.stylesheet = jss.createStyleSheet(styles as any);
  });

  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: `<t-button>Label</t-button>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should disable the button', () => {
    expect(element.disabled).toBe(false);
    element.disable();
    expect(element.disabled).toBe(true);
  });

  it('should enable the button', () => {
    element.disable();
    expect(element.disabled).toBe(true);
    element.enable();
    expect(element.disabled).toBe(false);
  });

  it('should handle size css class', () => {
    element.size = 'small';
    expect(element.getCssClassMap()).toContain(
      `${element.stylesheet.classes.button}--size-small`
    );
  });

  it('should handle variant css class', () => {
    element.variant = 'primary';
    expect(element.getCssClassMap()).toContain(
      `${element.stylesheet.classes.button}--variant-primary`
    );
  });

  it('should handle disabled css class', () => {
    element.disabled = true;
    expect(element.getCssClassMap()).toContain(
      `${element.stylesheet.classes.button}--disabled`
    );
  });

  it('should handle deselected css class', () => {
    element.deselected = true;
    expect(element.getCssClassMap()).toContain(
      `${element.stylesheet.classes.button}--deselected`
    );
  });
});
