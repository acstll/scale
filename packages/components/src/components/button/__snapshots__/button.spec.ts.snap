// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should match snapshot 1`] = `
<t-button>
  <mock:shadow-root>
    <style>
      .button-0-0-2 {
  color: #333;
  border: 1px solid #000;
  cursor: pointer;
  display: inline-flex;
  padding: 0.25rem 1rem;
  position: relative;
  font-size: 1rem;
  background: #fff;
  text-align: center;
  transition: all 0.2s ease-in-out;
  align-items: center;
  flex-shrink: 0;
  line-height: 2rem;
  user-select: none;
  border-radius: 4px;
  vertical-align: middle;
  justify-content: center;
}
.button-0-0-2:hover, .button-0-0-2.active {
  color: #fff;
  background: #000;
  transition: all 0.2s ease-in-out;
  border-color: #000;
  text-decoration: none;
}
.button-0-0-2:not(.tabbing):focus {
  outline: 0;
}
.button-0-0-2--disabled, .button-0-0-2--disabled:hover {
  color: #000;
  border: 1px solid #000;
  cursor: not-allowed;
  opacity: 0.5;
  background: #fff;
}
    </style>
    <button class="button-0-0-2">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Label
</t-button>
`;
